### PR TITLE
 User has to  login to see content

### PR DESCRIPTION
**Pull request by Noelryn**
### PT Link
https://www.pivotaltracker.com/story/show/161715906

### Included PR
* Adding before action to authenticate users in the app
* Adjusting feature tests 
* Adding ajax error message 
* Minor styling in the login form

### Screenshot
login

### CODE REVIEW

File: app/controllers/jobs_controller.rb [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # frozen_string_literal: true
 
 class JobsController < ApplicationController
+  before_action :authenticate_user!
```
Comment (tochman) [inline app/controllers/jobs_controller.rb]: If this is a general rule, then extract this to the ApplictionController.

Comment (rupertlion) [reply app/controllers/jobs_controller.rb]: At the moment given how a user can create another user (RSC creates RSC) I think this interferes with that process if inheriting the before action in the user controller. However, when we change to the devise_inviter method for that feature, inheriting from the application controller would be the way to go. At least that's IMO

Comment (Noelryn) [reply app/controllers/jobs_controller.rb]: We have resolved this

File: features/user_has_to_login_to_see_content.feature [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        And I click on "Logout" 
+
+    Scenario: User has successfully login and can see content of website
+        Given I fill in "Email" with "email@email.com"
+        And I fill in "Password" with "mypassword"
+        And I click on "Log in"
+        Then I should see "intern"
+       
+       Scenario: User tries to access content without logging in
+       Given I am on the landing page
```
Comment (Prozak8) [inline features/user_has_to_login_to_see_content.feature:24]: This step is not needed.

Edit: I read the background wrong.

File: app/javascript/css/tailwind.scss [hunk 3]
```diff
@@ -213,15 +213,15 @@ a {
 }
 
 .button {
-    @apply bg-grey-dark text-white ml-4 py-2 px-3 no-underline ;
+    @apply bg-grey-darker text-white font-bold py-2 px-5 no-underline ;
```
Comment (rupertlion) [inline app/javascript/css/tailwind.scss]: this is used elsewhere - please check and create a new class if required

File: app/javascript/css/tailwind.scss [hunk 4]
```diff
@@ -213,15 +213,15 @@ a {
...
 
     &:hover {
         @apply bg-grey-darker;
     }
 }
 
 .shared_devise_link {
-    @apply bg-grey-dark text-white ml-4 py-2 px-3 no-underline ;
+    @apply bg-grey-darker text-white font-bold py-2 px-5 no-underline ;
```
Comment (rupertlion) [inline app/javascript/css/tailwind.scss:228]: check this is not used in other views that require the original styling

File: app/views/devise/shared/_links.html.haml [hunk 5]
```diff
@@ -5,7 +5,7 @@
   = link_to "Sign up", new_registration_path(resource_name), remote: true, class: 'shared_devise_link',  onclick: 'sharedDeviseLinkHandler(event);'
   %br/
 - if devise_mapping.recoverable? && controller_name != 'passwords' && controller_name != 'registrations'
-  = link_to "Forgot your password?", new_password_path(resource_name), class: 'shared_devise_link'
+  = link_to "Forgot your password?", new_password_path(resource_name)
```
Comment (rupertlion) [inline app/views/devise/shared/_links.html.haml:8]: if you are not using the shared_devise_link class here, then need to check why we change it at all in the tailwind.scss file

File: features/user_has_to_login_to_see_content.feature [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    Background:
+        Given the following registered RSC exist
+            | email           | password   |
+            | email@email.com | mypassword |
+        And the user is logged in as "email@email.com"
+        And the following jobs exist in the database
+            | name     |
+            | intern   |
+        And I am on the landing page 
+        And I click on "Logout"
```
Comment (ESteinman) [inline features/user_has_to_login_to_see_content.feature:15]: What's the idea here? You're `logged in as` and then `logout` in the same background-step? Would you even have to log in, in the first place?

Comment (rupertlion) [reply features/user_has_to_login_to_see_content.feature:15]: I already spotted and we're on it ;-)

File: features/rsc_can_register_to_account.feature [hunk 7]
```diff
@@ -28,18 +28,17 @@ Feature: RSC creates a new rsc user
     Scenario: RSC logs out and try to log in with an incorrect email
         Given I click on "X"
         And I click on "Logout"
-        Then I should be on the landing page
-        And I click on "Login"
+        Then I should see "Log in" 
         And I fill in "Email" with "wrong@email.com"
         And I fill in "Password" with "password"
         And I click on "Log in"
+        # Then stop
```
Comment (Prozak8) [inline features/rsc_can_register_to_account.feature]: Why is this here?

Comment (Noelryn) [reply features/rsc_can_register_to_account.feature]: This has been removed, good spot

File: app/views/devise/sessions/_new.html.haml [hunk 8]
```diff
@@ -1,5 +1,4 @@
-- content_for :extras do
-  = button_tag 'X', id: 'cancel_modal', class: "cancel-button"
+
 - content_for :modal_content do
   = render partial: 'login_form'
```
Comment (RhonGabriel) [inline app/views/devise/sessions/_new.html.haml:4]: why the 'x' button has been removed?

Comment (Noelryn) [reply app/views/devise/sessions/_new.html.haml:4]: Because we don't need it on this modal, The point is that you need to log in to access the site so you shouldn't be able to cancel out of it

**Comment by Prozak8** [comment]
Please update the pivotal tracker so that the tasks reflect the description of this PR

File: app/javascript/css/tailwind.scss (hunk 3 above)
Comment (Noelryn) [reply app/javascript/css/tailwind.scss]: Changed this back to the previous code

File: app/javascript/css/tailwind.scss (hunk 4 above)
Comment (Noelryn) [reply app/javascript/css/tailwind.scss:228]: This is only currently being used with the log in and sign up buttons and doesnt effect anything else

File: app/views/devise/shared/_links.html.haml (hunk 5 above)
Comment (Noelryn) [reply app/views/devise/shared/_links.html.haml:8]: This was used on the forgot your password link and was bunching all the links together when it is removed from the fort password link, it looks a lot better.

**Review by Prozak8 — APPROVED** [review]

**Review by RhonGabriel — APPROVED** [review]

**Review by rupertlion — APPROVED** [review]

**Review by ESteinman — APPROVED** [review]
